specifyweb/frontend/js_src/lib/components/SchemaConfig/UniquenessRules.tsx

```diff
@@ -0,0 +1,674 @@
+                handleChanged({
+                  ...rule,
+                  fields: replaceItem(rule.fields, index, newField.name),
+                });
+              }}
+            />
+            {rule.fields.length > 1 && (
+              <Button.Icon
+                className={`w-fit ${className.dataEntryRemove}`}
+                disabled={readOnly}
+                icon="minus"
+                title={commonText.remove()}
+                onClick={(): void =>
+                  handleChanged({
+                    ...rule,
+                    fields: removeItem(rule.fields, index),
+                  })
+                }
+              />
+            )}
+          </div>
+        ))}
+        <Button.BorderedGray
+          className="w-fit"
+          disabled={readOnly}
+          onClick={(): void =>
+            handleChanged({
+              ...rule,
+              fields: insertItem(
+                rule.fields,
+                rule.fields.length,
+                fields.find(({ name }) => !rule.fields.includes(name))!.name
+              ),
+            })
+          }
+        >
+          {commonText.add()}
+        </Button.BorderedGray>
+        <p>{schemaText.scope()}</p>
+        <UniquenessRuleScope
+          model={model}
+          rule={rule}
+          onChange={handleChanged}
+        />
+        <Input.Text
+          disabled
+          value={
+            rule.scopes.length === 0
+              ? schemaText.database()
+              : getFieldsFromPath(model, rule.scopes[0])
+                  .map((field) => field.localization.name as string)
+                  .join(' -> ')
+          }
+        />
+      </>
+    </Dialog>
+  );
+}
+
+function UniquenessRuleScope({
+  rule,
+  model,
+  onChange: handleChanged,
+}: {
+  readonly rule: UniquenessRule;
+  readonly model: SpecifyModel;
+  readonly onChange: (newRule: typeof rule) => void;
+}): JSX.Element {
+  const [mappingPath, setMappingPath] = React.useState<RA<string>>(
+    rule.scopes.length === 0 ? ['database'] : rule.scopes[0].split('__')
+  );
+
+  const databaseScopeData: Readonly<Record<string, HtmlGeneratorFieldData>> = {
+    database: {
+      isDefault: true,
+      isEnabled: true,
+      isRelationship: false,
+      optionLabel: schemaText.database(),
+    },
+  };
+
+  const getValidScopeRelationships = (
+    model: SpecifyModel
+  ): Readonly<Record<string, HtmlGeneratorFieldData>> =>
+    Object.fromEntries(
+      model.relationships
+        .filter(
+          (relationship) =>
+            !(['one-to-many', 'many-to-many'] as RA<RelationshipType>).includes(
+              relationship.type
+            ) && !relationship.isVirtual
+        )
+        .map((relationship) => [
+          relationship.name,
+          {
+            isDefault: false,
+            isEnabled: true,
+            isRelationship: true,
+            optionLabel: relationship.localization.name as string,
```

Comment:
can `as string` be replaced with `!`?

Reply:
Fixed in https://github.com/specify/specify7/pull/4215/commits/43aee5b97e52edca55b58e0cb2b76773754b510f